internal/handler/common/update.go

```diff
@@ -31,28 +33,177 @@ import (
+//
+// In case of updating multiple documents, UpdateDocument returns an error immediately after one of the
+// operation fails. The rest of the documents are not processed.
+// TODO https://github.com/FerretDB/FerretDB/issues/2612
+func UpdateDocument(ctx context.Context, c backends.Collection, cmd string, iter types.DocumentsIterator, param *Update) (*UpdateResult, error) { //nolint:lll // for readability
+	result := new(UpdateResult)
 
-	if update.Len() == 0 {
-		// replace to empty doc
-		for _, key := range doc.Keys() {
-			docUpdated = true
+	isFindAndModify := (strings.ToLower(cmd) == "findandmodify")
+
+	for {
+		var upsert, modified bool
+
+		_, doc, err := iter.Next()
+		if err != nil {
+			if !errors.Is(err, iterator.ErrIteratorDone) {
+				return nil, lazyerrors.Error(err)
+			}
+
+			if result.Matched.Count == 0 && param.Upsert {
+				upsert = true
+				doc = must.NotFail(types.NewDocument())
+			}
+
+			if !upsert {
+				return result, nil
+			}
+		}
 
-			if key != "_id" {
-				doc.Remove(key)
+		if upsert {
+			if err = processFilterEqualityCondition(doc, param.Filter); err != nil {
+				return nil, lazyerrors.Error(err)
+			}
+		} else {
+			result.Matched.Count++
+			if isFindAndModify {
+				result.Matched.Doc = doc.DeepCopy()
 			}
 		}
 
-		return docUpdated, nil
+		if !param.HasUpdateOperators {
+			modified, err = processReplacementDoc(cmd, doc, param.Update)
+		} else {
+			modified, err = processUpdateOperator(cmd, doc, param.Update, upsert)
+		}
+
+		if err != nil {
+			return nil, lazyerrors.Error(err)
+		}
+
+		if !doc.Has("_id") {
+			doc.Set("_id", types.NewObjectID())
+		}
+
+		// TODO https://github.com/FerretDB/FerretDB/issues/3454
+		if err = doc.ValidateData(); err != nil {
+			return nil, lazyerrors.Error(err)
+		}
+
+		if upsert {
+			_, err = c.InsertAll(ctx, &backends.InsertAllParams{Docs: []*types.Document{doc}})
+			if err != nil {
+				return nil, lazyerrors.Error(err)
+			}
+			result.Upserted.Doc = doc
+
+			// upsert happens only once, no need to iterate further
+			return result, nil
+		} else if modified {
+			_, err := c.UpdateAll(ctx, &backends.UpdateAllParams{Docs: []*types.Document{doc}})
+			if err != nil {
+				return nil, lazyerrors.Error(err)
+			}
+
+			result.Modified.Count++
+			if isFindAndModify {
+				result.Modified.Doc = doc
+			}
+		}
 	}
+}
+
+// processFilterEqualityCondition copies the fields with equality condition from filter to doc.
+func processFilterEqualityCondition(doc, filter *types.Document) error {
+	iter := filter.Iterator()
+	defer iter.Close()
+
+	for {
+		key, val, err := iter.Next()
+		if err != nil {
+			if errors.Is(err, iterator.ErrIteratorDone) {
+				return nil
+			}
+
+			return lazyerrors.Error(err)
+		}
+
+		if key[0] == '$' {
+			continue
+		}
+
+		if valDoc, ok := val.(*types.Document); ok {
+			if valDoc.Len() == 1 {
+				valEq, _ := valDoc.Get("$eq")
+				if valEq == nil {
+					continue
```

Comment:
If `valDoc` doesn't contain `$eq` filter but instead contains a document without an operator i.e. `{"v", bson.D{{"k1", "v1"}}}`, should it `continue`? I think it needs to set val/valDoc to doc. A test would be great :hugs:

Reply:
Good catch. Fixed it, added a test case also.